### PR TITLE
A massive refactor on the Syntax Grammar for full token colorization. 

### PR DESCRIPTION
**Pull request by pouyakary**
Hey there. Thanks for the great plugin. I'm the author of the [Pro Colors](https://marketplace.visualstudio.com/items?itemName=karyfoundation.theme-karyfoundation-themes) themes. What I wanted to achieve with this theme was to have every token colored as they should be. This is an example of that in TypeScript:

Pro Colors on TypeScript
 
Every token is colorized. Now, I came across coding with Flutter that I realized the grammar has massive problems and that is not something that I liked:

Dart Language before the new patch

So I worked into the grammar and fixed and improved as many things as I could, while at the same time adding first-class support to Pro Colors. the result of that is this:

Dart code after the patch

# Changes

- So there are lots of things that got changed. First of all, I formatted the code for readability with [Virtuous](https://marketplace.visualstudio.com/items?itemName=karyfoundation.virtuous-vscode)

- `~/` used to be rendered as two separate operators, it's fixed now.

- class definitions added.

- function definitions added.

Screenshot 1401-04-31 at 18 38 29

- punctuations now have grammars, and can be colorized separately, and have the right priority in the rules.

Screenshot 1401-04-31 at 18 42 27

- every keyword and type now includes a `$1` in their definition, so for example, `keyword.control.dart` becomes `keyword.control.if.dart` which makes it possible for theme authors like me to customize every single word as we wish.

Screenshot 1401-04-31 at 18 42 54

- Type Parameters `<T>` are colorized in the right way:

Screenshot 1401-04-31 at 18 43 28

- Object Literal Keys are added:

Screenshot 1401-04-31 at 18 44 20

- fixed the rule name `"variable.other.source.dart` to `comment.block.documentation.dart`

- matching the `=>` and `(` after the function identifier

Screenshot 1401-04-31 at 18 45 57

- moving all the root patterns to the `block` repository (to use for string interpolation)

- moving `$` into the string interpolation variable

Screenshot 1401-04-31 at 18 46 44

- better string interpolation grammar

Screenshot 1401-04-31 at 18 47 42

Screenshot 1401-04-31 at 18 47 53

### CODE REVIEW

**Comment by pouyakary** [comment]
Added coloring of the object before the dot notation:

Screenshot 1401-05-01 at 20 28 19

**Comment by pouyakary** [comment]
The class definition has some problems, I'm going to work on that...

**Comment by DanTup** [comment]
Thanks for the contribution! I haven't had chance to look at this in detail yet, but I'll try to do so next week (although I may wait until you think you're done first).

I had been planning to look at this a little in the near future with a goal of making the textmate grammar more closely match the semantic tokens output (see #4062 for one example of why). We'll need to make sure that any changes made in this PR don't make that any worse (I suspect if anything they will make it better, but it's something we'll need to verify). I'd really like to have some better (automated) testing for that since right now changes to this file can be risky, but it's not something that exists today.

Once we are happy with this, we should send the PR to https://github.com/dart-lang/dart-syntax-highlight since that's really the official source of truth for this file now.

File: .vscode/settings.json [hunk 1]
```diff
@@ -1,34 +1,37 @@
...
+        "src/test/test_projects/flutter_create_basic",
+        "src/test/test_projects/flutter_create_sample"
+    ],
+    "dart.doNotFormat": [
+        "**/lib/comments.dart"
+    ],
+    "cSpell.words": [
+        "dartdoc",
+        "oldschool"
+    ]
```
Comment (DanTup) [inline .vscode/settings.json:36]: I think this file may have been included in error, it doesn't appear to contain any relevant changes.

Comment (pouyakary) [reply .vscode/settings.json:36]: I used the C-Spell plugin in vscode to check the spellings and these two words raised errors, so I added them to the workspace dictionary. You can delete it if you find it useless

**Comment by pouyakary** [comment]
So I worked a little bit on the class definitions and now it includes the inherited class as well:

Screenshot 1401-05-02 at 11 01 13

**Comment by pouyakary** [comment]
@DanTup thanks for checking it out. I actually used the JavaScript/TypeScript definitions as a reference (since they are the best grammars implemented so far) and then used their naming (thought well, most themes are targeting those so for example if I change `entity.other.inherited-class.js` to `.dart` it'll make the most themes work out of the box with Dart too. So hope it was not a bad break.)

However, Thanks a lot for the heads up! I'll be forking that then :)

**Comment by DanTup** [comment]
@pouyakary I had a quick compare of this versus the current grammar. Some items like `get` `set` as identifiers look better now (they're no longer coloured as keywords), however it looks like `String` has changed now to no longer be coloured like a class (left is before, right is after):

Screenshot 2022-07-25 at 11 51 37

While String is a built-in type, it's coloured the same as a class when using Semantic Tokens and I definitely don't want to introduce any changes that make additional differences to that (the fewer differences between the grammar and the semantic tokens, the less "flicker" there is when opening a new file when you see the grammar briefly and then the semantic tokens). For references, the same code using semantic tokens looks like this:

Screenshot 2022-07-25 at 11 52 03

I've also added pushed a change to master that now dumps the tokens to snapshot files (see c43941e2 and 65a577da) using the [vscode-tmgrammar-test](https://github.com/PanAeon/vscode-tmgrammar-test) package that will make verifying changes to the grammar a little easier (although the current test files are quite simple).

If you rebase on `master` you can run `npm run update-grammar-snapshots` which will re-generate those snapshots with your changes (which will be required for the tests to pass, allowing the changes to be reviewed in the PR). Feel free to add additional code to the files in the `grammar_testing` project if there are specific things missing you'd like to ensure don't regress.

(Ultimately I'd like to move tests like that to the dart-lang/syntax-highlight repo, but since this is already using TS/npm it was easier to try out here).

**Comment by pouyakary** [comment]
Fixed the problem with the class definition where the generics in the extending type did not color:

Screenshot 1401-05-04 at 17 37 39

**Comment by pouyakary** [comment]
@DanTup thanks a lot for your awesome investigation 👍🏻 Let me fix the String and get/set right away and go to the tests...

**Comment by pouyakary** [comment]
So I had counted Strings as primitive types and it seems that they are rendered correctly as primitive. But I can change the string to a class maybe? I don't know that, please guide me on what to expect of strings?

Screenshot 1401-05-04 at 17 48 14

**Comment by DanTup** [comment]
@pouyakary yep, I think we should treat `String` as a class. This matches the existing grammar and semantic tokens. I understand `String` is perhaps a little odd because it could be argued to belong to two categories (similar to `void`), but since a goal here is not to diverge further from the semantic tokens colouring, I think we should keep it as-is.

**Comment by pouyakary** [comment]
@DanTup sure thing, I'll fix it then.

**Comment by pouyakary** [comment]
Spread operator added (referencing JS as always)
 
Screenshot 1401-05-05 at 16 42 32

**Comment by pouyakary** [comment]
Added the optionality keyword:

Screenshot 1401-05-05 at 18 38 38

**Comment by DanTup** [comment]
Cool, thanks!

I've started some work to try and get some tests into the dart-syntax-highlighting repo, which should make it easier to verify the impact of these changes.

I think we should try to minimise the non-functional changes in this change to make it easier to review. For example the original file is indented with tabs, but this PR changes everything to spaces (which makes every line look like it was modified). If you can re-format using tabs, the diff will be much smaller and easier to review. Thanks!

**Comment by pouyakary** [comment]
Fixed the problem with nested generics

**Comment by DanTup** [comment]
> Fixed the problem with nested generics

Nice! I just discovered that issue recently while working on some testing support for this and hadn't gotten to looking at it yet.

I've made some steps towards having better automated tests using the DevTools parser (see https://github.com/flutter/devtools/commit/91cf832f9983698597f3030e02473643a73da930), although there are some issues with it right now (it doesn't produce the correct spans compared to other tools). I'd like to resolve those issues so we can then get tests also in the dart-syntax-highlighting repo, which will make it easier to accept larger changes like this without fear of regressing anything.

Are you able to align the indenting in this file with the original so the diff is just what's _functionally_ changed, and remove the changes to the vscode/settings file? That will make it easier to review once everything is ready.

Thanks!

**Comment by pouyakary** [comment]
@DanTup I actually can I guess, since every different atomic change has been a commit. So even the formatting is a commit. I don't know if you have seen the history of it, but can I ask you to do so? I have an initial commit that the original version formatted. And then on top of that I have committed new changes. so it's readable what is changed. However I can recommit all of those commits on the original version (although I find this one more readable).

But the thing is: If I'm going to recommit this all with the original format, lets as well do it in the upstream repo so that it contains the history. we can stop this PR here and move that all there?

**Comment by DanTup** [comment]
I understand, but it's far easier for me to review a single diff than go through all 28 commits and review them individually. I don't think changing the tabs to spaces is an improvement to readability. Tab sizes can easily be controlled in the editor but spaces cannot. Additionally, changing every lines indenting makes it much more difficult to track down regressions if they occur, because you have to step through more layers of history to find when a line was last modified.

Sending the PR to dart-syntax-highlighting is fine too. They're both essentially the same file and can be copied either way (from there to here or here to there). In either case, it'll need careful review and testing.

Thanks!

**Comment by pouyakary** [comment]
Okay then let me move to that repo and recommit these there with the proper style.

**Comment by pouyakary** [comment]
A new full grammar for the `get` / `set` pattern so that the next token after them are marked as a function definition (as what they are)

**Comment by pouyakary** [comment]
So I didn't know that the get could be async and even generator :) how cool is that :) anyhow, that is fixed:

Screenshot 1401-05-13 at 21 15 16
Screenshot 1401-05-13 at 21 15 25
Screenshot 1401-05-13 at 21 22 09

**Comment by pouyakary** [comment]
The extension definition grammar is added:

Screenshot 1401-05-14 at 22 35 30

Screenshot 1401-05-14 at 22 35 26

Screenshot 1401-05-14 at 22 35 35

Screenshot 1401-05-14 at 22 35 38

---

Screenshot 1401-05-14 at 22 35 05

Screenshot 1401-05-14 at 22 35 11

Screenshot 1401-05-14 at 22 35 17

**Comment by pouyakary** [comment]
Extension definition grammar is added:

Screenshot 1401-05-14 at 22 39 21
Screenshot 1401-05-14 at 22 39 25
Screenshot 1401-05-14 at 22 39 28
Screenshot 1401-05-14 at 22 39 31

---
Screenshot 1401-05-14 at 22 39 46

Screenshot 1401-05-14 at 22 39 52

Screenshot 1401-05-14 at 22 39 59

**Comment by pouyakary** [comment]
@DanTup sorry that I haven't had a chance to work on the tests. Still working on the grammar as I learn the language

**Comment by DanTup** [comment]
np, I haven't finished setting up testing for the dart-syntax-highlighting repo yet. But what we could do in the meantime is add more files in https://github.com/Dart-Code/Dart-Code/tree/master/src/test/test_projects/grammar_testing/lib that cover the cases you've been fixing and testing. We could add them and generate the snapshot files in another PR, and then rebase this and re-generate them. That way this diff would include both the grammar changes, and the changes to the snapshots, which will make it easier to review and confirm everything is working as expected.

**Comment by pouyakary** [comment]
The `extends` keyword didn't get color in the generics, that is fixed now:

Screenshot 1401-05-15 at 00 45 38

**Comment by pouyakary** [comment]
@DanTup awesome

**Comment by pouyakary** [comment]
I just noticed that the name in the extension definition is optional and so I made it so:

Screenshot 1401-05-15 at 15 26 28

**Comment by pouyakary** [comment]
Class names (in definition) can now also be lowercase.

Screenshot 1401-05-15 at 15 31 13

**Comment by pouyakary** [comment]
Added the generics after the destination:

Screenshot 1401-05-15 at 17 19 49

**Comment by pouyakary** [comment]
Just found a new problem:

Screen Shot 2022-08-13 at 5 43 53 PM

**Comment by DanTup** [comment]
@pouyakary sorry for the delay, there's been a lot of out-of-office over the Summer. I've finished getting https://github.com/dart-lang/dart-syntax-highlight/pull/28 ready for review, so now it's possible to add tests in the dart-syntax-highlight repository (at least, once the PR is landed).

I've scanned though your changes above and tried to ensure there is test code that covers the things you've fixed, but if I have miss any, please let me know (or prepare a PR that adds additional test code in `test/test_files`). This will make it easier to review changes to see the impact of them without lots of manual testing.

It might make sense to try and split your change up into a few separate reviews to make it easier to review (for example separating the bug fixes, the general improvements, and the new scopes added to previously unmarked tokens).

If anything is unclear, or you need anything from me, please do let me know!

**Comment by pouyakary** [comment]
@DanTup thanks a lot for that! I too encountered a lots of personal things that delayed my side too. I have found a few problems in the grammar. Some related to operators, some related to extensions and mostly the generics. I'll fix them soon and add the tests, thanks a lot for adding them. Hope this PR gets complete soon !

**Comment by pouyakary** [comment]
String is now a class and all classes can be customizable in the themes:

Screen Shot 2022-10-05 at 4 35 12 PM

**Comment by pouyakary** [comment]
The precedence of assignment operators are fixed so `+=` will no longer be colored as `+` and `=`

Screen Shot 2022-10-05 at 4 35 58 PM

**Comment by pouyakary** [comment]
I encountered a problem with a dart library that had uppercase function names. And the grammar counted them as classes:

Screen Shot 2022-09-22 at 11 29 42 AM

I have to play with the grammar to fix this but for now, I just fixed it enough to not recognize it as anything:

Screen Shot 2022-10-05 at 4 41 03 PM

**Comment by pouyakary** [comment]
There was a problem with function definition grammar where it counted await functionName() as definition. It's now resolved:

Screen Shot 2022-09-22 at 6 05 40 PM

**Comment by pouyakary** [comment]
The list of problems that I had found is now fully fixed. the next level would be for me to add tests

**Comment by pouyakary** [comment]
It would be cool if you could check the definitions too @DanTup

**Comment by pouyakary** [comment]
Bug fix. The `expand_less :` is recognized as a property key in this screen shot:

Screen Shot 2022-10-06 at 12 52 54 AM

And is fixed in this one:

Screen Shot 2022-10-06 at 5 57 44 PM

**Comment by pouyakary** [comment]
@DanTup don't be, I delayed as well. The formatting is because it's just really hard for me to work with the json and this way make it readable for me. I'll fix the spacing. And yeah I think it should be like that. however I have no idea how to move all the history of commits and comments to there.

For the automated testing. I got blocked on the look behind regrex features of the testing. and have no idea how to fix the grammars with that. If that could be solved sure.

And for the semantic tokens. I have used the TypeScript/JavaScript grammar of Microsoft/vscode to see how they view a grammar and since I think all major themes on vscode first implement the TS/JS grammar, it will have some solid grounds. The tokens that go over the bounds of textmate spec are all based on the namings of the JS. I don't know if that's a right idea but I had no better one.

**Comment by pouyakary** [comment]
Needless to say that for each token that you have a better scope I'll appreciate. My only wish doing this was to differentiate between the tokens and have all of them colored.

**Comment by DanTup** [comment]
> The formatting is because it's just really hard for me to work with the json and this way make it readable for me. I'll fix the spacing. And yeah I think it should be like that. however I have no idea how to move all the history of commits and comments to there.

I don't think it'll be easy to move all of the commits as-is to another repo, although depending on the scale of the overall changes, a single commit of the new file might not be too difficult to review.

However, you should remove any formatting changes, including changes like this that change the order of fields:

Screenshot 2022-10-11 at 14 20 00

Changes like this make the diff much larger and more difficult to review. Changes are likely to be reviewed and merged much sooner if they are easier to review, because they'll take less time for the reviewer.


> For the automated testing. I got blocked on the look behind regrex features of the testing. and have no idea how to fix the grammars with that. If that could be solved sure.

I'm not very familiar with the look behind pattern you used. Could you provide an example (with an explanation of why it's needed) and I can see whether I can find an alternative that works with the Dart regex? Thanks!


> Needless to say that for each token that you have a better scope I'll appreciate. My only wish doing this was to differentiate between the tokens and have all of them colored.

Yep, anything working towards this is a great change IMO. Our grammar/semantic tokens are currently quite different (many things are not coloured in the grammar) and it's something I've wanted to improve for a while :-)

**Comment by DanTup** [comment]
@pouyakary what's the current status of this? I replied on https://github.com/dart-lang/dart-syntax-highlight/issues/32 about the issue you had - I didn't really understand the special syntax being used but think we should stick to thinks that Dart can parse (since the grammar is used by more than just VS Code, for example Dart DevTools uses it and parses it using Dart).

**Comment by pouyakary** [comment]
@DanTup I'm actually using the grammar everyday and try to fix the bugs as I encounter them. (yesterday I found out dart can have generator getters and my grammar breaks at `get something sync* =>`, something I have to push a patch for soon) and then the testing remains which unfortunately I still haven't had the time to go through. And moving the commits to the main repository has become of a challenge for me. 

Let me make the patch right now and move the commits to the main dart repo. And since now I've written the most of the grammar, I think I have to request access to the repo and maintain it myself. (which I have no idea who to ask for that)

Let me do the patch now and make a PR today to actually ship a version of this.

**Comment by DanTup** [comment]
> And moving the commits to the main repository has become of a challenge for me.

I'm happy to help with this if I can, but we'll need to create test cases for the the things being changed so they can be reviewed (and ensure they don't regress). I think a good first step would be to just extend the tests in https://github.com/dart-lang/dart-syntax-highlight to include code that covers all the things you're changing in the golden tests (right now it will produce the wrong results, but this will make it easier to review the differences when committing the grammar afterwards).

>  And since now I've written the most of the grammar, I think I have to request access to the repo and maintain it myself. (which I have no idea who to ask for that)

Do you mean the https://github.com/dart-lang/dart-syntax-highlight repo? That's maintained by the Dart team and the best way to get changes is by opening pull requests (which is what I do when I have improvements to make). This grammar is used by many tools so its important all changes are carefully reviewed and tests added to prevent regressions.

Let me know if there's anything I can do to help. Thanks!

**Comment by pouyakary** [comment]
@DanTup okay sure. Let me gradually add things there and use this issue as the changelog.

**Pull request closed without merging**